server/src/main/java/io/druid/server/coordinator/CostBalancerStrategy.java

```diff
@@ -81,50 +73,107 @@ public ServerHolder findNewSegmentHomeBalancer(
     return chooseBestServer(proposalSegment, serverHolders, true).rhs;
   }
 
+  static double computeJointSegmentsCost(final DataSegment segment, final Iterable<DataSegment> segmentSet)
+  {
+    double totalCost = 0;
+    for(DataSegment s : segmentSet) {
+      totalCost += computeJointSegmentsCost(segment, s);
+    }
+    return totalCost;
+  }
+
   /**
-   * This defines the unnormalized cost function between two segments.  There is a base cost given by
-   * the minimum size of the two segments and additional penalties.
-   * recencyPenalty: it is more likely that recent segments will be queried together
+   * This defines the unnormalized cost function between two segments.
+   *
    * dataSourcePenalty: if two segments belong to the same data source, they are more likely to be involved
    * in the same queries
-   * gapPenalty: it is more likely that segments close together in time will be queried together
    *
-   * @param segment1 The first DataSegment.
-   * @param segment2 The second DataSegment.
+   * intervalPenalty: it is more likely that segments close together in time will be queried together
+   *
+   * @param segmentA The first DataSegment.
+   * @param segmentB The second DataSegment.
    *
    * @return The joint cost of placing the two DataSegments together on one node.
    */
-  public double computeJointSegmentCosts(final DataSegment segment1, final DataSegment segment2)
+  public static double computeJointSegmentsCost(final DataSegment segmentA, final DataSegment segmentB)
   {
-    final long gapMillis = gapMillis(segment1.getInterval(), segment2.getInterval());
+    final Interval intervalA = segmentA.getInterval();
+    final Interval intervalB = segmentB.getInterval();
 
-    final double baseCost = Math.min(segment1.getSize(), segment2.getSize());
-    double recencyPenalty = 1;
-    double dataSourcePenalty = 1;
-    double gapPenalty = 1;
+    final double t0 = intervalA.getStartMillis();
+    final double t1 = (intervalA.getEndMillis() - t0) / MILLIS_FACTOR;
+    final double start = (intervalB.getStartMillis() - t0) / MILLIS_FACTOR;
+    final double end = (intervalB.getEndMillis() - t0) / MILLIS_FACTOR;
 
-    if (segment1.getDataSource().equals(segment2.getDataSource())) {
-      dataSourcePenalty = 2;
-    }
+    final double multiplier = segmentA.getDataSource().equals(segmentB.getDataSource()) ? 2.0 : 1.0;
```

Comment:
There's an extra penalty to co-locating segments from the same source. The idea is that a given query hits a single datasource, so this encourages segments from the same source to be more spread out than segments from different sources (which are unlikely to be queried at the same time).